dot/parachain/backing/active_leaves_update.go

```diff
@@ -0,0 +1,528 @@
+		}
+
+		var mode parachaintypes.ProspectiveParachainsMode
+		leaf, ok := cb.perLeaf[maybeNewRP]
+		if !ok {
+			// If the relay-parent isn't a leaf itself,
+			// then it is guaranteed by the prospective parachains
+			// subsystem that it is an ancestor of a leaf which
+			// has prospective parachains enabled and that the
+			// block itself did.
+			mode = *prospectiveParachainsMode
+		} else {
+			mode = leaf.prospectiveParachainsMode
+		}
+
+		// construct a `PerRelayParent` from the runtime API and insert it.
+		rpState, err := constructPerRelayParentState(cb.BlockState, maybeNewRP, &cb.keystore, mode)
+		if err != nil {
+			return fmt.Errorf("constructing per relay parent state for relay-parent %s: %w", maybeNewRP, err)
+		}
+
+		if rpState != nil {
+			cb.perRelayParent[maybeNewRP] = rpState
+		}
+	}
+	return nil
+}
+
+func processRemainingSeconded(
+	cb *CandidateBacking,
+	remainingSeconded map[parachaintypes.CandidateHash]parachaintypes.ParaID,
+	leafHash common.Hash,
+) map[parachaintypes.ParaID]*btree.Map[uint, parachaintypes.CandidateHash] {
+	var wg sync.WaitGroup
+	var mut sync.Mutex
+	secondedAtDepth := make(map[parachaintypes.ParaID]*btree.Map[uint, parachaintypes.CandidateHash])
+
+	for candidateHash, parID := range remainingSeconded {
+		wg.Add(1)
+		go func(candidateHash parachaintypes.CandidateHash, parID parachaintypes.ParaID) {
+			defer wg.Done()
+
+			getTreeMembership := parachaintypes.ProspectiveParachainsMessageGetTreeMembership{
+				ParaID:        parID,
+				CandidateHash: candidateHash,
+				ResponseCh:    make(chan []parachaintypes.FragmentTreeMembership),
+			}
+
+			cb.SubSystemToOverseer <- getTreeMembership
+			membership := <-getTreeMembership.ResponseCh
+
+			for _, m := range membership {
+				if m.RelayParent == leafHash {
+					mut.Lock()
+
+					tree, ok := secondedAtDepth[parID]
+					if !ok {
+						tree = new(btree.Map[uint, parachaintypes.CandidateHash])
+					}
+
+					for _, depth := range m.Depths {
+						tree.Load(depth, candidateHash)
+					}
+					mut.Unlock()
+				}
+			}
+		}(candidateHash, parID)
+	}
+	wg.Wait()
+	return secondedAtDepth
+}
+
+func cleanUpPerRelayParentByLeafAncestry(cb *CandidateBacking) {
+	remaining := make(map[common.Hash]bool)
+
+	for hash := range cb.perLeaf {
+		remaining[hash] = true
+	}
+
+	allowedRelayParents := cb.implicitView.allAllowedRelayParents()
+	for _, relayParent := range allowedRelayParents {
+		remaining[relayParent] = true
+	}
+
+	keysToDelete := []common.Hash{}
+	for rp := range cb.perRelayParent {
+		if _, ok := remaining[rp]; !ok {
+			keysToDelete = append(keysToDelete, rp)
+		}
+	}
+
+	for _, key := range keysToDelete {
+		delete(cb.perRelayParent, key)
+	}
+}
+
+func removeUnknownRelayParentsFromPerCandidate(cb *CandidateBacking) {
+	keysToDelete := []parachaintypes.CandidateHash{}
+
+	for candidateHash, pc := range cb.perCandidate {
+		if _, ok := cb.perRelayParent[pc.relayParent]; !ok {
+			keysToDelete = append(keysToDelete, candidateHash)
```

Comment:
same as the previous comment,

It's deleting the entry from the same map we loop over.
So, if we delete here inside the loop, it might lead to unexpected behaviour.